// mongoose-express-sanitizer/dist/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _helpers = require("./helpers");

var general = function general(schemaTree, data) {
  return sanitize(schemaTree, data);
};

var sanitize = function sanitize(schemaTree, data) {
  Object.keys(schemaTree).forEach(function (key) {
    var attribute = schemaTree[key];

    if ((0, _typeof2["default"])(attribute) === 'object' && attribute !== null) {
      if (attribute.required && !data[key]) {
        throw {
          message: "The attribute ".concat(key, " is required."),
          httpErrorCode: 400
        };
      }

      if (data[key]) {
        if ((0, _typeof2["default"])(attribute) === 'object' && (0, _typeof2["default"])(data[key]) === 'object' && !Array.isArray(data[key])) {
          sanitize(attribute, data[key]);
        } else {
          (0, _helpers.validateElementOfSchema)(attribute, data[key]);
        }
      }
    }
  });
};

var headers = function headers(schemaTree, req) {
  return sanitize(schemaTree, req.headers);
};

var queries = function queries(schemaTree, req) {
  return sanitize(schemaTree, req.queries);
};

var body = function body(schemaTree, req) {
  return sanitize(schemaTree, req.body);
};

var params = function params(schemaTree, req) {
  return sanitize(schemaTree, req.params);
};

module.exports = {
  general: general,
  headers: headers,
  queries: queries,
  body: body,
  params: params
};
var _default = general;
exports["default"] = _default;
